rig/machine_control/machine_controller.py

```diff
@@ -1657,51 +1659,120 @@ def get_p2p_routing_table(self, x, y):
+        info = self._send_scp(x, y, 0, SCPCommands.info, expected_args=3)
+        return ChipInfo(
+            num_cores=info.arg1 & 0x1F,
+            core_states=[consts.AppState(c)
+                         for c in struct.unpack("<18B", info.data)],
+            working_links=set(link for link in Links
+                              if (info.arg1 >> (8 + link)) & 1),
+            largest_free_sdram_block=info.arg2,
+            largest_free_sram_block=info.arg3,
+        )
+
+    @ContextMixin.use_contextual_arguments()
     def get_working_links(self, x, y):
         """Return the set of links reported as working.
 
-        The returned set lists only links over-which nearest neighbour
-        peek/poke commands could be sent. This means that links connected to
-        peripherals may falsely be omitted.
+        This command tests each of the links leaving a chip by sending a PEEK
+        nearest-neighbour packet down each link to verify that the remote
+        device is a SpiNNaker chip. If no reply is received via a given link or
+        if the remote device is not a SpiNNaker chip, the link is reported as
+        dead.
+
+        See also: :py:meth:`.get_chip_info`.
 
         Returns
         -------
         set([:py:class:`rig.machine.Links`, ...])
         """
-        link_up = self.read_struct_field("sv", "link_up", x, y)
-        return set(link for link in Links if link_up & (1 << link))
+        return self.get_chip_info(x, y).working_links
 
     @ContextMixin.use_contextual_arguments()
     def get_num_working_cores(self, x, y):
-        """Return the number of working cores, including the monitor."""
+        """Return the number of working cores, including the monitor.
+
+        See also: :py:meth:`.get_chip_info`.
+        """
         return self.read_struct_field("sv", "num_cpus", x, y)
 
-    def get_machine(self, x=0, y=0, default_num_cores=18):
+    def _get_all_chips_info(self, x=0, y=0):
+        """For internal use only. Get system dimensions and a ChipInfo for all
+        working chips.
+
+        Returns
+        -------
+        (width, height, chips_info)
+            width and height are integers giving the dimensions of the system.
+
+            chips_info is a dictionary {(x, y): :py:class:`.ChipInfo`, ...}
+            giving the chip info for all chips which are reported as working in
+            the P2P table and which responded to :py:meth:`.get_chip_info`.
+        """
+        # A quick way of getting a list of working chips
+        p2p_tables = self.get_p2p_routing_table(x, y)
+
+        # Calculate the extent of the system
+        max_x = max(x_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+        max_y = max(y_ for (x_, y_), r in iteritems(p2p_tables)
+                    if r != consts.P2PTableEntry.none)
+
+        chips_info = {}
+
+        for (x, y), p2p_route in iteritems(p2p_tables):
+            if x <= max_x and y <= max_y:
```

Comment:
This line is a throwback from the darker days when `get_p2p_routing_table` actually grabbed the whole table rather than just the portion for the machine which actually exists. As such it should, of course, be removed but first I need to work out why it doesn't fail the branch coverage checker...